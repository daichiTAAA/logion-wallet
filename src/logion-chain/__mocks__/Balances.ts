import { getBalances } from './BalancesMock'

export { getBalances }
