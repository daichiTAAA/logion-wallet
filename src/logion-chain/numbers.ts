export function amount(balance: string, decimals: number): string {
    const prefixed = new ScientificNumber(balance, 0);
    const converted = prefixed.convertTo(decimals);
    return converted.unnormalize();
}

export function balance(amount: string, decimals: number): string {
    const prefixed = new ScientificNumber(amount, decimals);
    const converted = prefixed.convertTo(0);
    return converted.unnormalize();
}

export class ScientificNumber {
    private _normalized: string;
    private _tenExponant: number;

    constructor(coefficient: string, tenExponant: number) {
        this._normalized = this.normalize(coefficient);
        this._tenExponant = tenExponant;
    }

    private shiftDecimalSeparator(positions: number): string {
        if(positions === 0) {
            return this._normalized;
        } else {
            const { integerPart, decimalPart } = this.split();
            if(positions > 0) {
                let rightWithDot;
                if(decimalPart.length <= positions) {
                    rightWithDot = decimalPart.padEnd(decimalPart.length + positions, "0");
                    rightWithDot = rightWithDot + ".";
                } else  { // rightPart.length > positions
                    const pivot = positions;
                    rightWithDot = decimalPart.substring(0, pivot) + "." + decimalPart.substring(pivot);
                }
                return integerPart + rightWithDot;
            } else { // positions < 0
                let leftWithDot;
                if(integerPart.length <= -positions) {
                    leftWithDot = integerPart.padStart(integerPart.length - positions - 1, "0");
                    leftWithDot = "." + leftWithDot;
                } else  { // leftPart.length > -positions
                    const pivot = integerPart.length + positions;
                    leftWithDot = integerPart.substring(0, pivot) + "." + integerPart.substring(pivot);
                }
                return this.zeroStripRight(leftWithDot + decimalPart);
            }
        }
    }

    private split() {
        const dotPosition = this._normalized.indexOf('.');
        const integerPart = this._normalized.substring(0, dotPosition);
        const decimalPart = this._normalized.substring(dotPosition + 1);
        return {
            integerPart,
            decimalPart
        };
    }

    private normalize(num: string): string {
        let sanitized = this.zeroStripLeft(num);
        if(sanitized.indexOf('.') === -1) {
            return sanitized + ".";
        } else {
            return this.zeroStripRight(sanitized);
        }
    }

    zeroStripLeft(num: string): string {
        return num.replace(/^0*/, '');
    }

    zeroStripRight(num: string): string {
        return num.replace(/0*$/, '');
    }

    unnormalize(): string {
        if(this._normalized === ".") {
            return "0";
        } else if(this._normalized.startsWith(".")) {
            return "0" + this._normalized;
        } else if(this._normalized.endsWith(".")) {
            return this._normalized.substring(0, this._normalized.length - 1);
        } else {
            return this._normalized;
        }
    }

    convertTo(newTenExponent: number): ScientificNumber {
        const offset = this._tenExponant - newTenExponent;
        return new ScientificNumber(this.shiftDecimalSeparator(offset), newTenExponent);
    }

    get normalized(): string {
        return this._normalized;
    }

    get tenExponant(): number {
        return this._tenExponant;
    }

    optimizeScale(maxDigits: number): ScientificNumber {
        const { integerPart, decimalPart } = this.split();
        if(integerPart.length > maxDigits) {
            return this.limitIntegerDigits(integerPart, maxDigits);
        } else {
            const relevantDecimalPart = this.zeroStripLeft(decimalPart);
            if(decimalPart.length > relevantDecimalPart.length) {
                return this.limitDecimalDigits(decimalPart, relevantDecimalPart);
            } else {
                return this;
            }
        }
    }

    limitIntegerDigits(integerPart: string, maxDigits: number): ScientificNumber {
        const delta = integerPart.length - maxDigits;
        return this.convertTo(this._tenExponant + delta);
    }

    limitDecimalDigits(decimalPart: string, relevantDecimalPart: string): ScientificNumber {
        const delta = relevantDecimalPart.length - decimalPart.length - 1;
        return this.convertTo(this._tenExponant + delta);
    }
}

export class PrefixedNumber {
    private _scientificNumber: ScientificNumber;
    private _prefix: UnitPrefix;

    constructor(num: string, prefix: UnitPrefix) {
        this._scientificNumber = new ScientificNumber(num, prefix.tenExponant);
        this._prefix = prefix;
    }

    convertTo(prefix: UnitPrefix): PrefixedNumber {
        const newScaled = this._scientificNumber.convertTo(prefix.tenExponant);
        return new PrefixedNumber(newScaled.normalized, prefix);
    }

    get normalized(): string {
        return this._scientificNumber.normalized;
    }

    get prefix(): UnitPrefix {
        return this._prefix;
    }

    optimizeScale(maxDigits: number): PrefixedNumber {
        const optimized = this._scientificNumber.optimizeScale(maxDigits);
        const optimalPrefix = closestPrefix(optimized.tenExponant);
        return this.convertTo(optimalPrefix);
    }
}

export const EXA: UnitPrefix = {
    symbol: 'E',
    tenExponant: 18,
};

export const PETA: UnitPrefix = {
    symbol: 'P',
    tenExponant: 15,
};

export const TERA: UnitPrefix = {
    symbol: 'T',
    tenExponant: 12,
};

export const GIGA: UnitPrefix = {
    symbol: 'G',
    tenExponant: 9,
};

export const MEGA: UnitPrefix = {
    symbol: 'M',
    tenExponant: 6,
};

export const KILO: UnitPrefix = {
    symbol: 'k',
    tenExponant: 3,
};

export const NONE: UnitPrefix = {
    symbol: '',
    tenExponant: 0,
};

export const MILLI: UnitPrefix = {
    symbol: 'm',
    tenExponant: -3,
};

export const MICRO: UnitPrefix = {
    symbol: 'Âµ',
    tenExponant: -6,
};

export const NANO: UnitPrefix = {
    symbol: 'n',
    tenExponant: -9,
};

export const PICO: UnitPrefix = {
    symbol: 'p',
    tenExponant: -12,
};

export const FEMTO: UnitPrefix = {
    symbol: 'f',
    tenExponant: -15,
};

export const ATTO: UnitPrefix = {
    symbol: 'a',
    tenExponant: -18,
};

export interface UnitPrefix {
    symbol: string,
    tenExponant: number,
}

const SORTED_UNITS: UnitPrefix[] = [
    EXA,
    PETA,
    TERA,
    GIGA,
    MEGA,
    KILO,
    NONE,
    MILLI,
    MICRO,
    NANO,
    PICO,
    FEMTO,
    ATTO
].sort((a, b) => a.tenExponant - b.tenExponant);

function closestPrefix(tenExponant: number): UnitPrefix {
    let from: number = 0;
    let to: number = SORTED_UNITS.length;
    let candidate: number = -1;
    while((to - from) > 1) {
        const next = Math.floor(from + (to - from) / 2);
        if(SORTED_UNITS[next].tenExponant > tenExponant) {
            to = next;
        } else if(SORTED_UNITS[next].tenExponant < tenExponant) {
            from = next;
        } else {
            candidate = next;
            break;
        }
    }
    if((to - from) <= 1) {
        candidate = from;
    }
    if(candidate === 0) {
        const candidateDistance = Math.abs(SORTED_UNITS[candidate].tenExponant - tenExponant);
        const nextDistance = Math.abs(SORTED_UNITS[candidate + 1].tenExponant - tenExponant);
        if(nextDistance < candidateDistance) {
            return SORTED_UNITS[candidate + 1];
        } else {
            return SORTED_UNITS[candidate];
        }
    } else if(candidate === SORTED_UNITS.length - 1) {
        const candidateDistance = Math.abs(SORTED_UNITS[candidate].tenExponant - tenExponant);
        const previousDistance = Math.abs(SORTED_UNITS[candidate - 1].tenExponant - tenExponant);
        if(previousDistance < candidateDistance) {
            return SORTED_UNITS[candidate - 1];
        } else {
            return SORTED_UNITS[candidate];
        }
    } else {
        const candidateDistance = Math.abs(SORTED_UNITS[candidate].tenExponant - tenExponant);
        const previousDistance = Math.abs(SORTED_UNITS[candidate - 1].tenExponant - tenExponant);
        const nextDistance = Math.abs(SORTED_UNITS[candidate + 1].tenExponant - tenExponant);
        if(previousDistance < candidateDistance && previousDistance < nextDistance) {
            return SORTED_UNITS[candidate - 1];
        } else if(nextDistance < candidateDistance && nextDistance < previousDistance) {
            return SORTED_UNITS[candidate + 1];
        } else {
            return SORTED_UNITS[candidate];
        }
    }
}