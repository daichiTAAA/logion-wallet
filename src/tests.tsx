import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
