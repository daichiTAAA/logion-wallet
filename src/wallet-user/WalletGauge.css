.WalletGauge .actions {
    display: flex;
    justify-content: center;
}

.WalletGauge.arc .actions {
    margin-top: 60px;
    margin-bottom: 30px;
    margin-left: auto;
    margin-right: auto;
}

.WalletGauge .actions button:nth-child(n+2) {
    margin-left: 25px;
}

.WalletGauge.linear {
    display: flex;
    align-items: flex-end;
}

.WalletGauge.linear .actions {
    margin-bottom: 10px;
}
