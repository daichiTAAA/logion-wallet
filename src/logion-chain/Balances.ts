import { ApiPromise } from '@polkadot/api';
import { PrefixedNumber, ScientificNumber, convertToPrefixed, NONE, ATTO } from "./numbers";
import { ExtrinsicSubmissionParameters, Unsubscriber, signAndSend } from './Signature';
import { Call } from "@polkadot/types/interfaces";
import { SubmittableExtrinsic } from "@polkadot/api/submittable/types";

const LGNT_DECIMALS = 18;
export const LGNT_SMALLEST_UNIT = ATTO;
export const SYMBOL = "LGNT";

export interface GetAccountDataParameters {
    api: ApiPromise,
    accountId: string,
}

export interface AccountData {
    available: string,
    reserved: string,
    total: string,
}

export async function getAccountData(parameters: GetAccountDataParameters): Promise<AccountData> {
    const {
        api,
        accountId,
    } = parameters;

    const accountData = await api.query.system.account(accountId);

    return {
        available: accountData.data.free.toString(),
        reserved: accountData.data.reserved.toString(),
        total: accountData.data.free.add(accountData.data.reserved).toString(),
    };
}
export interface Coin {
    id: string,
    name: string,
    iconId: string,
    iconType: 'svg' | 'png',
    symbol: string,
}

export interface CoinBalance {
    coin: Coin,
    balance: PrefixedNumber,
    level: number,
}

export async function getBalances(parameters: GetAccountDataParameters): Promise<CoinBalance[]> {
    const {
        api,
        accountId,
    } = parameters;

    const data = await getAccountData({
        api: api!,
        accountId,
    });

    const logAvailable = scientificLogBalance(data.total).optimizeScale(3);
    const logPrefixedAvailable = convertToPrefixed(logAvailable);
    const logLevel = logAvailable.divideBy(ARTIFICIAL_MAX_BALANCE).toNumber();

    return [
        buildCoinBalance('lgnt', logPrefixedAvailable, logLevel),
        buildCoinBalance('dot', new PrefixedNumber("0", NONE), 1)
    ];
}

export function scientificLogBalance(tokens: string): ScientificNumber {
    return new ScientificNumber(tokens, -LGNT_DECIMALS).optimizeScale(3);
}

export function prefixedLogBalance(tokens: string): PrefixedNumber {
    const scientific = scientificLogBalance(tokens);
    return convertToPrefixed(scientific);
}

const ARTIFICIAL_MAX_BALANCE = scientificLogBalance("100");

function buildCoinBalance(coinId: string, balance: PrefixedNumber, level: number): CoinBalance {
    const coin = getCoin(coinId);
    return {
        coin,
        balance,
        level,
    }
}

export function getCoin(coinId: string): Coin {
    if(coinId === 'dot') {
        return {
            id: 'dot',
            name: 'Polkadot',
            iconId: 'dot',
            iconType: 'png',
            symbol: 'DOT',
        };
    } else if(coinId === "lgnt") {
        return {
            id: 'lgnt',
            name: 'logion Token',
            iconId: 'lgnt',
            iconType: 'svg',
            symbol: SYMBOL,
        };
    } else {
        throw new Error(`Unsupported coin ${coinId}`);
    }
}

export interface TransferParameters extends ExtrinsicSubmissionParameters, BuildTransferCallParameters {
}

export function transfer(parameters: TransferParameters): Unsubscriber {
    const {
        signerId,
        callback,
        errorCallback,
    } = parameters;

    return signAndSend({
        signerId,
        submittable: transferSubmittable(parameters),
        callback,
        errorCallback,
    });
}

export interface BuildTransferCallParameters {
    api: ApiPromise;
    destination: string;
    amount: PrefixedNumber;
}

function transferSubmittable(parameters: BuildTransferCallParameters): SubmittableExtrinsic<'promise'> {
    const {
        api,
        destination,
        amount
    } = parameters;
    return api.tx.balances.transfer(destination, amount.convertTo(LGNT_SMALLEST_UNIT).coefficient.unnormalize())
}

export function buildTransferCall(parameters: BuildTransferCallParameters): Call {
    return parameters.api.createType('Call', transferSubmittable(parameters))
}

export async function estimateFee(parameters: BuildTransferCallParameters): Promise<PrefixedNumber> {
    const submittable = transferSubmittable(parameters);
    const paymentInfo = await submittable.paymentInfo(parameters.destination);
    return new PrefixedNumber(paymentInfo.partialFee.toString(), LGNT_SMALLEST_UNIT);
}
