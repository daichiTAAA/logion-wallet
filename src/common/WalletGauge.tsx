import { useState, useCallback, useEffect } from 'react';
import { ATTO, FEMTO, MICRO, MILLI, NANO, NONE, PICO, PrefixedNumber } from '../logion-chain/numbers';
import { Coin, transfer } from '../logion-chain/Balances';

import Gauge from './Gauge';
import Button from './Button';
import Icon from './Icon';

import './WalletGauge.css';
import Dialog from './Dialog';
import FormGroup from './FormGroup';
import { useCommonContext } from './CommonContext';
import { Form, Spinner, InputGroup, DropdownButton, Dropdown } from 'react-bootstrap';
import { useLogionChain } from '../logion-chain';
import { isValidAccountId } from '../logion-chain/Accounts';
import ExtrinsicSubmitter, { SignAndSubmit } from '../ExtrinsicSubmitter';
import Alert from './Alert';

export interface Props {
    coin: Coin,
    balance: PrefixedNumber,
    level: number,
    type: 'arc' | 'linear',
}

interface State {
    status: Status;
    minExpectedTransactions?: number;
    interval?: number;
}

enum Status {
    IDLE,
    TRANSFERRING,
    EXPECTING_NEW_TRANSACTION,
    WAITING_FOR_NEW_TRANSACTION
}

export default function WalletGauge(props: Props) {
    const { colorTheme, accounts, refresh, transactions } = useCommonContext();
    const { api } = useLogionChain();
    const [ destination, setDestination ] = useState("");
    const [ amount, setAmount ] = useState("");
    const [ unit, setUnit ] = useState(NONE);
    const [ signAndSubmit, setSignAndSubmit ] = useState<SignAndSubmit>(null);
    const [ state, setState ] = useState<State>({status: Status.IDLE});

    const transferCallback = useCallback(() => {
        const signAndSubmit: SignAndSubmit = (setResult, setError) => transfer({
            api: api!,
            signerId: accounts!.current!.address,
            callback: setResult,
            errorCallback: setError,
            destination,
            amount: new PrefixedNumber(amount, unit),
        });
        setSignAndSubmit(() => signAndSubmit);
    }, [ accounts, amount, api, destination, unit ]);

    const successCallback = useCallback(() => {
        setDestination("");
        setAmount("");
        setSignAndSubmit(null);

        setState({
            status: Status.EXPECTING_NEW_TRANSACTION,
            minExpectedTransactions: transactions!.length + 1
        });
    }, [ transactions, setState ]);

    const cancelCallback = useCallback(() => {
        if(state.status === Status.WAITING_FOR_NEW_TRANSACTION) {
            window.clearInterval(state.interval);
        }
        setState({ status: Status.IDLE });
    }, [ state, setState ]);

    useEffect(() => {
        if(state.status === Status.EXPECTING_NEW_TRANSACTION) {
            const interval = window.setInterval(() => {
                refresh(false);
            }, 3000);

            setState({
                ...state,
                status: Status.WAITING_FOR_NEW_TRANSACTION,
                interval
            });
        }
    }, [ state, setState, refresh ]);

    useEffect(() => {
        if(state.status === Status.WAITING_FOR_NEW_TRANSACTION
                && transactions !== undefined
                && transactions!.length >= state.minExpectedTransactions!) {
            window.clearInterval(state.interval);
            setState({status: Status.IDLE});
        }
    }, [ state, setState, transactions ]);

    return (
        <div className={ "WalletGauge " + props.type }>
            <Gauge
                readingIntegerPart={ props.balance.coefficient.toInteger() }
                readingDecimalPart={ props.balance.coefficient.toFixedPrecisionDecimals(2) }
                unit={ props.balance.prefix.symbol + "LOG" }
                level={ props.level }
                type={ props.type }
            />
            <div className="actions">
                <Button slim onClick={() => setState({status: Status.TRANSFERRING})}><Icon icon={{id:'send'}} /> Send</Button>
            </div>
            <Dialog
                show={ state.status !== Status.IDLE }
                actions={[
                    {
                        id: 'cancel',
                        buttonText: "Cancel",
                        buttonVariant: 'secondary',
                        callback: cancelCallback
                    },
                    {
                        id: 'transfer',
                        buttonText: "Transfer",
                        buttonVariant: 'primary',
                        callback: transferCallback,
                        disabled: state.status !== Status.TRANSFERRING || !isValidAccountId(api!, destination) || isNaN(Number(amount)) || destination === accounts!.current!.address
                    }
                ]}
                size="lg"
            >
                <h3>Transfer LOGs</h3>
                <FormGroup
                    id="destination"
                    label="Destination"
                    control={<Form.Control
                        isInvalid={destination !== "" && (!isValidAccountId(api!, destination) || destination === accounts!.current!.address)}
                        type="text"
                        placeholder="The beneficiary's SS58 address"
                        value={destination}
                        onChange={value => setDestination(value.target.value)}
                    />}
                    colors={colorTheme.dialog}
                />
                <FormGroup
                    id="amout"
                    label="Amount"
                    noFeedback={true}
                    control={
                        <InputGroup hasValidation>
                            <Form.Control
                                isInvalid={amount !== "" && isNaN(Number(amount))}
                                type="text"
                                placeholder="The amount to transfer"
                                value={amount}
                                onChange={value => setAmount(value.target.value)}
                            />
                            <DropdownButton
                                as={InputGroup.Append}
                                title={ `${unit.symbol}LOG` }
                                id="input-group-dropdown-1"
                            >{
                                [
                                    NONE,
                                    MILLI,
                                    MICRO,
                                    NANO,
                                    PICO,
                                    FEMTO,
                                    ATTO
                                ].map(unit => <Dropdown.Item key={ unit.symbol } onClick={() => setUnit(unit)}>{ `${unit.symbol}LOG` }</Dropdown.Item>)
                            }</DropdownButton>
                            <Form.Control.Feedback type="invalid">
                                Please choose a username.
                            </Form.Control.Feedback>
                        </InputGroup>
                    }
                    colors={colorTheme.dialog}
                />
                <ExtrinsicSubmitter
                    id="transferLogs"
                    successMessage="Transfer successful."
                    signAndSubmit={ signAndSubmit }
                    onSuccess={ successCallback }
                    onError={ () => {} }
                />
                {
                    (state.status === Status.EXPECTING_NEW_TRANSACTION || state.status === Status.WAITING_FOR_NEW_TRANSACTION) &&
                    <Alert variant="info">
                        <Spinner animation="border"/>
                        <p>Transfer successful, waiting for the transaction to be finalized.</p>
                    </Alert>
                }
            </Dialog>
        </div>
    );
}
