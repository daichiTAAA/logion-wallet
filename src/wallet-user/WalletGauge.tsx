import { PrefixedNumber } from '../logion-chain/numbers';
import { Coin } from '../logion-chain/Balances';

import Gauge from '../common/Gauge';
import Button from '../common/Button';
import Icon from '../common/Icon';
import { ColorTheme } from '../common/ColorTheme';

import './WalletGauge.css';

export interface Props {
    coin: Coin,
    balance: PrefixedNumber,
    colorTheme: ColorTheme,
    level: number,
    type: 'arc' | 'linear',
}

export default function WalletGauge(props: Props) {

    return (
        <div className={ "WalletGauge " + props.type }>
            <Gauge
                readingIntegerPart={ props.balance.coefficient.toInteger() }
                readingDecimalPart={ props.balance.coefficient.toFixedPrecisionDecimals(2) }
                unit={ props.balance.prefix.symbol + "LOG" }
                level={ props.level }
                type={ props.type }
            />
            <div className="actions">
                <Button colors={ props.colorTheme.buttons } slim><Icon icon={{id:'send'}} /> Send</Button>
                <Button colors={ props.colorTheme.buttons } slim><Icon icon={{id:'buy'}} /> Buy</Button>
            </div>
        </div>
    );
}
