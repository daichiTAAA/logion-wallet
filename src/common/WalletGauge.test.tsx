jest.mock("../logion-chain");
jest.unmock("@logion/node-api/dist/Balances");

import { shallowRender } from '../tests';
import { getCoin } from "@logion/node-api/dist/Balances";
import { PrefixedNumber, MILLI } from '@logion/node-api/dist/numbers';

import WalletGauge from './WalletGauge';

test("renders arc", () => {
    testWalletGauge('arc');
});

function testWalletGauge(type: 'arc' | 'linear') {
    const result = shallowRender(<WalletGauge
        coin={ getCoin('lgnt') }
        balance={ new PrefixedNumber("20.00", MILLI) }
        level={ 0.5 }
        type={ type }
    />);
    expect(result).toMatchSnapshot();
}

test("renders linear", () => {
    testWalletGauge('linear');
});
