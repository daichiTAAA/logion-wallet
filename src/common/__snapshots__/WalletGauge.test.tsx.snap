// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders arc 1`] = `
<div
  className="WalletGauge arc"
>
  <Gauge
    level={0.5}
    readingDecimalPart="00"
    readingIntegerPart="20"
    type="arc"
    unit="mLOG"
  />
  <div
    className="actions"
  >
    <Button
      onClick={[Function]}
      slim={true}
    >
      <Icon
        icon={
          Object {
            "id": "send",
          }
        }
      />
       Send
    </Button>
  </div>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary",
          "callback": [Function],
          "id": "cancel",
        },
        Object {
          "buttonText": "Transfer",
          "buttonVariant": "primary",
          "callback": [Function],
          "disabled": true,
          "id": "transfer",
        },
      ]
    }
    show={false}
    size="lg"
  >
    <h3>
      Transfer LOGs
    </h3>
    <FormGroup
      colors={
        Object {
          "background": "#0c163d",
          "borderColor": "#e6007a",
          "foreground": "#ffffff",
        }
      }
      control={
        <FormControl
          isInvalid={false}
          onChange={[Function]}
          placeholder="The beneficiary's SS58 address"
          type="text"
          value=""
        />
      }
      id="destination"
      label="Destination"
    />
    <FormGroup
      colors={
        Object {
          "background": "#0c163d",
          "borderColor": "#e6007a",
          "foreground": "#ffffff",
        }
      }
      control={
        <InputGroup
          hasValidation={true}
        >
          <FormControl
            isInvalid={false}
            onChange={[Function]}
            placeholder="The amount to transfer"
            type="text"
            value=""
          />
          <DropdownButton
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "render": [Function],
              }
            }
            id="input-group-dropdown-1"
            title="LOG"
          >
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              LOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              mLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              µLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              nLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              pLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              fLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              aLOG
            </DropdownItem>
          </DropdownButton>
          <Feedback
            type="invalid"
          >
            Please choose a username.
          </Feedback>
        </InputGroup>
      }
      id="amout"
      label="Amount"
      noFeedback={true}
    />
    <ExtrinsicSubmitter
      id="transferLogs"
      onError={[Function]}
      onSuccess={[Function]}
      signAndSubmit={null}
      successMessage="Transfer successful."
    />
  </Dialog>
</div>
`;

exports[`renders linear 1`] = `
<div
  className="WalletGauge linear"
>
  <Gauge
    level={0.5}
    readingDecimalPart="00"
    readingIntegerPart="20"
    type="linear"
    unit="mLOG"
  />
  <div
    className="actions"
  >
    <Button
      onClick={[Function]}
      slim={true}
    >
      <Icon
        icon={
          Object {
            "id": "send",
          }
        }
      />
       Send
    </Button>
  </div>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary",
          "callback": [Function],
          "id": "cancel",
        },
        Object {
          "buttonText": "Transfer",
          "buttonVariant": "primary",
          "callback": [Function],
          "disabled": true,
          "id": "transfer",
        },
      ]
    }
    show={false}
    size="lg"
  >
    <h3>
      Transfer LOGs
    </h3>
    <FormGroup
      colors={
        Object {
          "background": "#0c163d",
          "borderColor": "#e6007a",
          "foreground": "#ffffff",
        }
      }
      control={
        <FormControl
          isInvalid={false}
          onChange={[Function]}
          placeholder="The beneficiary's SS58 address"
          type="text"
          value=""
        />
      }
      id="destination"
      label="Destination"
    />
    <FormGroup
      colors={
        Object {
          "background": "#0c163d",
          "borderColor": "#e6007a",
          "foreground": "#ffffff",
        }
      }
      control={
        <InputGroup
          hasValidation={true}
        >
          <FormControl
            isInvalid={false}
            onChange={[Function]}
            placeholder="The amount to transfer"
            type="text"
            value=""
          />
          <DropdownButton
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "render": [Function],
              }
            }
            id="input-group-dropdown-1"
            title="LOG"
          >
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              LOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              mLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              µLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              nLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              pLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              fLOG
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              aLOG
            </DropdownItem>
          </DropdownButton>
          <Feedback
            type="invalid"
          >
            Please choose a username.
          </Feedback>
        </InputGroup>
      }
      id="amout"
      label="Amount"
      noFeedback={true}
    />
    <ExtrinsicSubmitter
      id="transferLogs"
      onError={[Function]}
      onSuccess={[Function]}
      signAndSubmit={null}
      successMessage="Transfer successful."
    />
  </Dialog>
</div>
`;
